src/eval.c

```diff
@@ -3605,6 +3618,294 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+is_closing_quote(char_u *template, int is_literal_string, int *is_skipping_needed)
+{
+    return is_literal_string
+	? is_expected_closing('\'', '\'', '\'', template, is_skipping_needed)
+	: is_expected_closing('"', '\\', '"', template, is_skipping_needed);
+}
+
+/*
+ * Allocate a variable for $"${var}" and $'${val}' constant.
+ * Return OK or FAIL.
+ */
+    static int
+get_template_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+{
+    garray_T	plain;  // a string that is not a template string
+    char_u	*orig_arg = *arg;
+    int		is_literal_string = (*(*arg + 1) == '\'');
+    char_u	quote = is_literal_string ? '\'' : '"';
+    int		is_skipping_needed;
+
+    *arg += 2;  // $'
+
+    ga_init2(&plain, 1, 80);
+    ga_append(&plain, quote);
+
+    // Continue while it is not NULL and it is not a closing quote.
+    for (; (**arg != NUL) && !is_closing_quote(*arg, is_literal_string, &is_skipping_needed); ++*arg)
+    {
+	if (is_skipping_needed)
+	{
+	    ga_append(&plain, (int) **arg);
+	    ++*arg;
+	    ga_append(&plain, (int) **arg);
+	    continue;
+	}
+	else if (**arg == '$' && *(*arg + 1) == '$')
+	{
+	    // An escape of '$'
+	    ++*arg;
+	    ga_append(&plain, (int) **arg);
+	    continue;
+	}
+	else if (**arg == '$')
+	{
+	    char_u *stringified = (*(*arg + 1) == '{')
+		? enclose_template_expr_by_stringify(arg, is_literal_string)
+		: enclose_template_var_by_stringify(arg);
+	    if (stringified == NULL)
+	    {
+		ga_clear(&plain);
+		return FAIL;
+	    }
+
+	    ga_append(&plain, quote);            // '
+	    ga_concat(&plain, (char_u *) "..");  // ..
+	    ga_concat(&plain, stringified);      // stringify(x)
+	    ga_concat(&plain, (char_u *) "..");  // ..
+	    ga_append(&plain, quote);            // '
+
+	    vim_free(stringified);
+	    continue;
+	}
+	ga_append(&plain, (int) **arg);
+    }
+
+    if (**arg == NUL)
+    {
+	ga_clear(&plain);
+	semsg(_("E115: Missing quote: %s"), orig_arg);
+	return FAIL;
+    }
+    MB_PTR_ADV(*arg);
+    ga_append(&plain, quote);
+
+    if (!evaluate)
+    {
+	ga_clear(&plain);
+	return OK;
+    }
+    else
+    {
+	int	result;
+	char_u	*to_free = (char_u *) plain.ga_data;
+
+	result = eval1((char_u**) &plain.ga_data, rettv, TRUE);
+	vim_free(to_free);
+	return result;
+    }
+}
+
+    static int
+did_encount_string_quote(char_u **expr, int is_literal_string)
+{
+    // encounted a double quote in the single quoted string (lit str),
+    int did_encount_double_in_single =
+	(is_literal_string && **expr == '"');
+
+    // an escaped single quote in the single quoted string,
+    int did_encount_single_single_in_single =
+	(is_literal_string && **expr == '\'' && *(*expr + 1) == '\'');
+
+    // a single quote in the double quoted string,
+    int did_encount_single_in_double =
+	(!is_literal_string && **expr == '\'');
+
+    // or a single quote in the double quoted string?
+    int did_encount_escaped_double_in_double =
+	(!is_literal_string && **expr == '\\' && *(*expr + 1) == '"');
```

Comment:
I would use `is_escaped_quote()` also here. Then this line can be merged with above `did_encount_single_single_in_single` line.

Reply:
I merged the two var into `did_encount_escaped_quote` :+1:

```c
    static int
did_encount_string_quote(char_u **expr, int is_literal_string)
{
    // encounted a double quote in the single quoted string (lit str),
    int did_encount_double_in_single =
	(is_literal_string && **expr == '"');

    // a single quote in the double quoted string,
    int did_encount_single_in_double =
	(!is_literal_string && **expr == '\'');

    // or a single quote in the double quoted string?
    int did_encount_escaped_quote = is_escaped_quote(is_literal_string, *expr);

    int result =
	did_encount_double_in_single ||
	did_encount_single_in_double ||
	did_encount_escaped_quote;

    if (did_encount_escaped_quote)
    {
	++*expr;
    }

    return result;
}
```